### ISSUE TITLE
[PATCH] document + improve ancient optimization in ParseXS

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14329]
Migrated from [rt.perl.org#123418](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123418) (status was 'resolved')

Searchable as RT123418$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



See attached patch\. The actual optimization is wider than the mentioned 
types because of the regex\, but since nobody fixed the bug in 20 years\, 
it is probably not a bug and let us turn it into a feature\.

<details><summary>Perl Info</summary>

```

Flags:
               category=core
               severity=low

Site configuration information for perl 5.21.7:

Configured by Owner at Sat Nov 22 21:54:54 2014.

Summary of my perl5 (revision 5 version 21 subversion 7) configuration:
             Local Commit: 1bce52df028aabe28c20b2d97949e35c17ea811e
             Ancestor: 7072da8afeba4c87ae623cd913e274396ffcf1cd
             Platform:
               osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
               uname=''
               config_args='undef'
               hint=recommended, useposix=true, d_sigaction=undef
               useithreads=define, usemultiplicity=define
               use64bitint=undef, use64bitall=undef, uselongdouble=undef
               usemymalloc=n, bincompat5005=undef
             Compiler:
               cc='cl', ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG
-G7 -GL
-DWIN32 -D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS
-DPERL_HASH_FUNC_ONE_AT_A_TIME -DNO_MATHOMS -DPERL_IMPLICIT_CONTEXT
-DPERL_IMPLICIT_SYS -DUSE_PERLIO -D_USE_32BIT_TIME_T',
               optimize='-O1 -MD -Zi -DNDEBUG -G7 -GL',
               cppflags='-DWIN32'
               ccversion='13.10.6030', gccversion='', gccosandvers=''
               intsize=4, longsize=4, ptrsize=4, doublesize=8,
byteorder=1234,
doublekind=3
               d_longlong=undef, longlongsize=8, d_longdbl=define,
longdblsize=8,
longdblkind=0
               ivtype='long', ivsize=4, nvtype='double', nvsize=8,
Off_t='__int64',
lseeksize=8
               alignbytes=8, prototype=define
             Linker and Libraries:
               ld='link', ldflags ='-nologo -nodefaultlib -debug 
-opt:ref,icf
-ltcg  -libpath:"c:\perl521\lib\CORE"  -machine:x86'
               libpth="C:\Program Files\Microsoft Visual Studio .NET
2003\VC7\lib"
               libs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
               perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
               libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl521.lib
               gnulibc_version=''
             Dynamic Linking:
               dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
               cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf -ltcg  -libpath:"c:\perl521\lib\CORE"  -machine:x86'

Locally applied patches:
               ce7a4d57d0acca9f39a84d36d708c4505dfe45ca
               ca0b263f4b167ddf97416f657d79ab5bd3344357
               08919bf863666074243240abbd19cd1a74cc7b74
               b8a043377dbf39548709b107a11e5cc2714c0e9a
               efa855eb5cffb7739616c295dd968d1510efeeb0
               1d47d0b810e26d9a2f9101fb813bd5b3dd332cc9
               3faca062ddb056db54f73fa55b0a9d473675dd33
               0b3e905bda3e75ad948a1213f620656b60807393
               1b1efc719fde05d215e5a13fb38c03e12a3aab08
               1bce52df028aabe28c20b2d97949e35c17ea811e


@INC for perl 5.21.7:
               ..\lib
               C:/perl521/srcnewb4opt/lib
               .


Environment for perl 5.21.7:
               HOME (unset)
               LANG (unset)
               LANGUAGE (unset)
               LD_LIBRARY_PATH (unset)
               LOGDIR (unset)
               PATH=C:\WINDOWS\system32;C:\Program Files\Microsoft Visual
Studio
.NET 2003\Vc7\bin;C:\Program Files\Microsoft Visual Studio .NET
2003\Common7\IDE;C:\WINDOWS;C:\Program Files\Git\cmd;C:\Program
Files\Microsoft Visual Studio .NET 2003\Common7\Tools\bin;C:\perl\bin
               PERL_BADLANG (unset)
               PERL_JSON_BACKEND=Cpanel::JSON::XS
               PERL_YAML_BACKEND=YAML
               SHELL (unset)














```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-document-improve-ancient-optimization-in-ParseXS.patch</summary>

```diff
From df6e7e7d0c5bc57c288c3e48f4d5d234a584461f Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Fri, 12 Dec 2014 05:16:50 -0500
Subject: [PATCH] document + improve ancient optimization in ParseXS

From commit 8990e30710 "perl 5.0 alpha 6". The DESTROY XSUB with T_PTROBJ
as INPUT arg was refusing to obey my custom typemap for T_PTROBJ, but all
other XSUBs followed the custom T_PTROBJ override. Extend this
optimization to other rarely used typemap types and document this behavior.
---
 dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm |    5 ++++-
 dist/ExtUtils-ParseXS/lib/perlxstypemap.pod   |   15 +++++++++++++--
 2 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm b/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
index 8c7b09d..fedc494 100644
--- a/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
+++ b/dist/ExtUtils-ParseXS/lib/ExtUtils/ParseXS.pm
@@ -1867,7 +1867,10 @@ sub generate_init {
 
   my $typem = $typemaps->get_typemap(ctype => $type);
   my $xstype = $typem->xstype;
-  $xstype =~ s/OBJ$/REF/ if $self->{func_name} =~ /DESTROY$/;
+  #this is an optimization from perl 5.0 alpha 6, class check is skipped
+  #T_REF_IV_REF is missing since it has no untyped analog at the moment
+  $xstype =~ s/OBJ$/REF/ || $xstype =~ s/^T_REF_IV_PTR$/T_PTRREF/
+    if $self->{func_name} =~ /DESTROY$/;
   if ($xstype eq 'T_PV' and exists $self->{lengthof}->{$var}) {
     print "\t$var" unless $printed_name;
     print " = ($type)SvPV($arg, STRLEN_length_of_$var);\n";
diff --git a/dist/ExtUtils-ParseXS/lib/perlxstypemap.pod b/dist/ExtUtils-ParseXS/lib/perlxstypemap.pod
index 5b1c639..7d1f73c 100644
--- a/dist/ExtUtils-ParseXS/lib/perlxstypemap.pod
+++ b/dist/ExtUtils-ParseXS/lib/perlxstypemap.pod
@@ -469,6 +469,9 @@ The pointer is blessed into a class that is derived from the name
 of type of the pointer but with all '*' in the name replaced with
 'Ptr'.
 
+For C<DESTROY> XSUBs only, a T_PTROBJ is optimized to a T_PTRREF. This means
+the class check is skipped.
+
 =item T_REF_IV_REF
 
 NOT YET
@@ -477,12 +480,16 @@ NOT YET
 
 Similar to T_PTROBJ in that the pointer is blessed into a scalar object.
 The difference is that when the object is passed back into XS it must be
-of the correct type (inheritance is not supported).
+of the correct type (inheritance is not supported) while T_PTROBJ supports
+inheritance.
 
 The pointer is blessed into a class that is derived from the name
 of type of the pointer but with all '*' in the name replaced with
 'Ptr'.
 
+For C<DESTROY> XSUBs only, a T_REF_IV_PTR is optimized to a T_PTRREF. This
+means the class check is skipped.
+
 =item T_PTRDESC
 
 NOT YET
@@ -498,7 +505,11 @@ are no known users in core or on CPAN.
 
 =item T_REFOBJ
 
-NOT YET
+Like T_REFREF, except it does strict type checking (inheritance is not
+supported).
+
+For C<DESTROY> XSUBs only, a T_REFOBJ is optimized to a T_REFREF. This means
+the class check is skipped.
 
 =item T_OPAQUEPTR
 
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri Dec 12 02&#8203;:19&#8203;:41 2014\, bulk88 wrote&#8203;:
> This is a bug report for perl from bulk88@&#8203;hotmail\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> See attached patch\. The actual optimization is wider than the mentioned 
> types because of the regex\, but since nobody fixed the bug in 20 years\, 
> it is probably not a bug and let us turn it into a feature\.

Thanks\, applied as 3f6e436ff0889330b44f87823bf8f712a6da6aba\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
